sql: panic: cannot apply collation to non-string type

```
CREATE TABLE t (d DECIMAL);

ALTER TABLE t ALTER COLUMN d SET DATA TYPE NAME COLLATE d;
```

```
panic: cannot apply collation to non-string type: name [recovered]
        panic: panic while executing 1 statements: ALTER TABLE _ ALTER COLUMN _ SET DATA TYPE NAME COLLATE d; caused by cannot apply collation to non-string type: name

goroutine 3663 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc000ca3b00, 0x456b740, 0xc00231a100, 0x39d6860, 0xc0004b6660)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc000ca3b00, 0x456b740, 0xc00231a100)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x39d6860, 0xc0004b6660)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/types.MakeCollatedString(0x68490a0, 0xc0023b02e6, 0x1, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/types/types.go:600 +0x174
github.com/cockroachdb/cockroach/pkg/sql.applyColumnMutation(0xc002368000, 0xc001f58000, 0x4549100, 0xc001e58440, 0x456b740, 0xc001e58a80, 0xc001892240, 0xc000ca3e50, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/alter_table.go:744 +0x38c
github.com/cockroachdb/cockroach/pkg/sql.(*alterTableNode).startExec(0xc0002f1660, 0x456b740, 0xc001e58a80, 0xc001892240, 0xc000ca3e50, 0xc001c7a360, 0xc001e35a80)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/alter_table.go:565 +0x33d9
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2(0x3cb251b, 0xb, 0x456c800, 0xc0002f1660, 0xc0003d9260, 0xc001e35a90)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:500 +0x55
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1(0xc0012d6600, 0x3cb251b, 0xb, 0x456c800, 0xc0002f1660)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:142 +0x5d
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc0012d6600, 0x456c800, 0xc0002f1660, 0x3cb251b, 0xb)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:672 +0x286
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc0012d6600, 0x456c800, 0xc0002f1660, 0x0, 0x60)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:109 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:73
github.com/cockroachdb/cockroach/pkg/sql.startExec(0x456b740, 0xc001e58a80, 0xc001892240, 0xc000ca3e50, 0x456c800, 0xc0002f1660, 0x0, 0x6b59600)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:503 +0x286
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc000a5cf00, 0x456b740, 0xc001e58a80, 0x6b59600, 0x37470e0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:117 +0xd6
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc000a5cf00, 0x456b740, 0xc001e58a80)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:792 +0x52
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc001c7a360, 0x456b800, 0xc00195fb00, 0x3e27db8, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:584 +0x1ec
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0001048c0, 0xc0012d6540, 0xc0009c8090, 0xc001e36870, 0xc00223a840, 0xc000ca3f28, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:333 +0x381
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc0001048c0, 0x456b800, 0xc00195f080, 0xc000ca3f28, 0xc0012d6540, 0xc0009c8090, 0x456c800, 0xc0002f1660, 0xc00223a840, 0xc000ca3cd8)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:942 +0x20a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc000ca3b00, 0x456b800, 0xc00195f080, 0xc000ca3e50, 0x1, 0x7f735ba7b978, 0xc0023a6668, 0xc001e36a00, 0x0, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:812 +0x370
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc000ca3b00, 0x456b800, 0xc00195f080, 0xc000ca3e50, 0x7f735ba7b978, 0xc0023a6668, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:700 +0x75b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc000ca3b00, 0x456b800, 0xc00195f080, 0x4576080, 0xc00195ef00, 0xc0023b02ae, 0x39, 0x0, 0x1, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:416 +0xb64
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc000ca3b00, 0x456b800, 0xc00195f080, 0x4576080, 0xc00195ef00, 0xc0023b02ae, 0x39, 0x0, 0x1, 0x0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:98 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc000ca3b00, 0x456b740, 0xc001e58480, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1211 +0x1aba
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc000ca3b00, 0x456b740, 0xc00231a100, 0xc000520580, 0x5400, 0x15000, 0xc000520618, 0xc00154dd50, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1140 +0x1a3
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0001e3200, 0x456b740, 0xc00231a100, 0xc000ca3b00, 0x5400, 0x15000, 0xc000520618, 0xc00154dd50, 0x0, 0x0)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc0002ddd9f, 0xc0014c82c0, 0x456b740, 0xc00231a100, 0xc00154dd50, 0xc0001e3200, 0xc0023a6000, 0x4575f80, 0xc0014c8040, 0xc00182c9c0, ...)
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:584 +0x21c
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
        /home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:513 +0x17b
```

Assigning to y'all because this has to do with schema change. Assign back to me if it's really a SQL exec bug.

--- a/pkg/sql/types/types.go
+++ b/pkg/sql/types/types.go
@@ -593,7 +593,7 @@ func MakeQChar(width int32) *T {
 //
 func MakeCollatedString(strType *T, locale string) *T {
 	switch strType.Oid() {
-	case oid.T_text, oid.T_varchar, oid.T_bpchar, oid.T_char:
+	case oid.T_text, oid.T_varchar, oid.T_bpchar, oid.T_char, oid.T_name:
 		return &T{InternalType: InternalType{
 			Family: CollatedStringFamily, Oid: strType.Oid(), Width: strType.Width(), Locale: &locale}}
 	}
@@ -1480,7 +1480,11 @@ func (t *T) upgradeType() error {
 		t.InternalType.ArrayContents = Oid
 
 	case name:
-		t.InternalType.Family = StringFamily
+		if t.InternalType.Locale != nil {
+			t.InternalType.Family = CollatedStringFamily
+		} else {
+			t.InternalType.Family = StringFamily
+		}
 		t.InternalType.Oid = oid.T_name
 		if t.Width() != 0 {
 			return errors.AssertionFailedf("name type cannot have non-zero width: %d", t.Width())

--- a/pkg/sql/types/types_test.go
+++ b/pkg/sql/types/types_test.go
@@ -117,6 +117,9 @@ func TestTypes(t *testing.T) {
 		{MakeCollatedString(MakeQChar(20), enCollate),
 			MakeScalar(CollatedStringFamily, oid.T_char, 0, 20, enCollate)},
 
+		{MakeCollatedString(Name, enCollate), &T{InternalType: InternalType{
+			Family: CollatedStringFamily, Oid: oid.T_name, Locale: &enCollate}}},
+
 		// DATE
 		{Date, &T{InternalType: InternalType{
 			Family: DateFamily, Oid: oid.T_date, Locale: &emptyLocale}}},

